Enable PHP 8.0 Unit-Tests

nupplaphil opened the pull request:
This will enable PHP8.0 Unit-Testing.
- Using PHPUnit 8 (manually installed)
- Fixing all failed tests because of PHP8
- Fixing all failed tests because of PHPUnit 8

All changes are either in the `php.yaml` file of the github workflow or in test cases, so no impact on Friendica itself

A nice side-effect: PHPUnit8 reduces the duration of the unit-tests significant.

nupplaphil committed:
Adapt tests for PhpUnit 7.5

nupplaphil committed:
Enable PHP 8.0 Auto-Test

MrPetovan commented on composer.json:
```diff
@@ -13,7 +13,7 @@
 		"issues": "https://github.com/friendica/friendica/issues"
 	},
 	"require": {
-		"php": ">=7.0",
```
What feature made you bump this?

nupplaphil replied:
https://packagist.org/packages/phpunit/phpunit#7.5.0 <-- this one :)

MrPetovan replied:
Why do we need to upgrade to this version?

nupplaphil replied:
https://phpunit.de/supported-versions.html

Because even 7.5.0 is EOL and doesn't "officially" supports PHP 8.0 but at least supports PHP 7.1 (and worked with our tests)
The next step would be PHPUnit 8, which starts with PHP 7.2 but would officially support PHP 8.0

MrPetovan replied:
We can have dev dependencies that are based on the PHP executable version and normal dependencies that are locked in the lowest version necessary (currently 7.0). The CI just needs to run `composer update --ignore-platform-reqs` to get the latest PHPUnit while the rest of the libraries will be bound to their own version limits. This way we can even use PHPUnit 9 and Friendica itself will keep working with PHP 7.0+.

MrPetovan replied:
Even better, the CI can run `composer install --no-dev` and then `composer update phpunit/phpunit --ignore-platform-reqs` to limit the number of package updates.

nupplaphil replied:
So that means that we support more versions than we're currently are able to test? Sounds a little bit risky and could lead to unseen problems.
=> But I'll do so :-)

nupplaphil replied:
ok at least it's the same like we're already doing because we start with the CI with 7.2 .. so nevermind, you're right! :)

MrPetovan replied:
There are two levels of support: the technical level where the codebase *can* run on PHP 7.0+. And there's the official level of support at the lowest actively supported PHP version, currently it's 7.3+. This means that we reserve the right to discard any support requests on PHP versions 7.0-7.2 even though Friendica can work on them.

So we should drop the tests on version 7.2 and open them for version 8.0.

nupplaphil replied:
Puh .. I'll have to investigate a little bit more for this setup.

Because installing PHPunit without limitation of the platform will use all the "new" php 8.0 things for PHPunit as well -> This breaks our testcases, because they're inheriting from these PHP classes and now are incompatible (f.e. they're missing the new return types).

nupplaphil replied:
reverted platform minimum requirements => solved, see https://github.com/friendica/friendica/pull/10112#discussion_r605983300 for followup :D

nupplaphil commented:
OK, didn't found all test-execution problems ;-)

nupplaphil committed:
Fix tests
- Fix InstallerTest case
- Fix native function mocking
- Add returntype for Unit-Tests
- Fixing ExtendedPDO test troubles
- Fix wrong class inheritance path for DatabaseTestTrait.php
- Fix SyslogLogger Server Exception for PHP8
- Add user/contact to database.fixture.php
- Avoid invalid rollbacks for test setup in PHP8

nupplaphil commented on composer.json:
```diff
@@ -94,7 +94,7 @@
 	"config": {
 		"platform": {
-			"php": "7.0"
+			"php": "7.2"
```
@MrPetovan  - This line does the trick, see https://getcomposer.org/doc/06-config.md#platform
I can now assume using PHP 7.2 for installing dependencies, but the minimum requirements are still PHP 7.0.

Therefore I can require PHP-Unit with `--ignore-platform-reqs` into the `composer.json` and pin it at `composer.lock`, but there's still no failure during install. The only thing is that when installing the whole dependencies, the unit-tests are now just usable for PHP 7.2+

MrPetovan replied:
This is not good either. Since Friendica *can* run on PHP 7.0, the dependencies *should* also work. This means setting the platform value to the same as the minimum PHP requirement value so that we pull the correct dependency versions.

With your change, we could pull a PHP 7.2+ dependency version which would break Friendica for PHP 7.0-7.1 even though Friendica's own code could work on PHP 7.0.

The dev dependencies do not need to be bound to this version though, so using `--ignore-platform-reqs` is warranted to update PHPUnit.

MrPetovan replied:
I'm finally seeing your comment about backward-breaking changes in PHPUnit which I didn't think about. In this case it is better to use the PHPUnit version compatible with PHP 7.2 and lock that version in the `composer.lock` file for use with all the test environments. You can temporarily change the PHP version in `platform` to do this.

nupplaphil replied:
OK found another way => I'm manually installing the PHPUnit 8 phar per CI and so I don't need to adapt the composer files  and routines at all :-)

nupplaphil commented:
Finished, ready for review :-)

MrPetovan commented on .github/workflows/php.yml:
```diff
@@ -28,7 +28,7 @@ jobs:
       fail-fast: false
       matrix:
-        php-versions: ['7.2', '7.3', '7.4']
+        php-versions: ['7.2', '7.3', '7.4', '8.0']
```
As per our conversations about supported versions, please remove unsupported version 7.2.
```suggestion
        php-versions: ['7.3', '7.4', '8.0']
```

nupplaphil replied:
I don't think we should remove 7.2
- It's supported for PHPUnit 8
- It doesn't slow down the overal test
- The more tests we can have for the current technical supported PHP version, the better

MrPetovan replied:
It's symbolic, we only officially support PHP versions that still are supported by the PHP Group. It isn't a question of performance.

MrPetovan commented on .github/workflows/php.yml:
```diff
@@ -62,6 +62,11 @@ jobs:
       - name: Install dependencies
         run: composer install --prefer-dist
 
+      - name: Install PHPUnit 8
+        run: curl -O -L https://phar.phpunit.de/phpunit-8.phar
+          && chmod +x phpunit-8.phar
+          && mv phpunit-8.phar /usr/local/bin/phpunit
```
What about running tests locally? Do we also need to install PHPUnit separately?

nupplaphil replied:
There's no difference to the current situation.

But I think you're right that because of "johnkary/phpunit-speedtrap", we pin phpunit at 5.8 in `composer.lock`
Shall I remove this dependency, so the phpunit pinning is deleted as well?

I think the speedtrap isn't necessary anymore because the speed of just the phpunit tests is <1m for PHP8 and ~2-3m for PHP7

nupplaphil replied:
Ah, and to answer your question: Yes we do need to install phpunit separately, but for me it feels more transparent than force-upgrading a current phpunit version per `--ignore-platform-req`

Shall I add it to the `README.md` ?

MrPetovan replied:
Yes please, and remove the testing dependencies from the composer.json

nupplaphil replied:
I created a setup-script instead of manually executing the command and added it to a new `README.md` inside the `tests/` directory.

nupplaphil committed:
Remove 7.2 PHP Unit-Test support

nupplaphil committed:
Remove PHPUnit dependency & create shell-script with description for it

MrPetovan commented:
Thank you for your work!
